clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -200,6 +200,41 @@ bool Sema::isKnownGoodSYCLDecl(const Decl *D) {
   return false;
 }
 
+bool isZeroSizedArray(QualType Ty) {
+  if (const auto *CATy = dyn_cast<ConstantArrayType>(Ty)) {
+    return (CATy->getSize() == 0);
+  }
+  return false;
+}
+
+void Sema::checkSYCLVarDeclIfInKernel(VarDecl *Var) {
+  // not all variable types supported in kernel contexts
+  // if not we record a deferred diagnostic.
+  assert(getLangOpts().SYCLIsDevice &&
+         "Should only be called during SYCL compilation");
+  QualType Ty = Var->getType();
+  SourceRange Loc = Var->getLocation();
+
+  // __int128, __int128_t, __uint128_t
+  if (Ty->isSpecificBuiltinType(BuiltinType::Int128) ||
+      Ty->isSpecificBuiltinType(BuiltinType::UInt128))
+    SYCLDiagIfDeviceCode(Loc.getBegin(), diag::err_type_unsupported)
+        << Ty.getUnqualifiedType().getCanonicalType().getAsString();
```

Comment:
You shouldn't have to do 'getAsString', the << operator accepts Type and QualType.